### PR TITLE
Typescript

### PR DESCRIPTION
**Pull request by maximilianMairinger**
TLDR: Read this [TS quickstart guide](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html). Run 'npm run build' every time before publishing. For development, run `npm run dev` and which executes repl/src/repl.ts in the browser. `npm run nodeDev` does the same for node development. Testing procedure stays unchanged.

-----------

I implemented the build infrastructure I usually use, which is a bit more extensive than what is absolutely necessary and may seem overwhelming at first, but I'll explain everything new in detail below.

I noticed you already used a comment based typing system. TS does basically the same thing, the only difference being that it's native to the language. Meaning, you need to compile (transpile to be exact) it back to JS to remove the type annotations before using it again in the browser or node (deno would support it natively). Similar to your typing system, it is also optional, thus all valid JS is valid TS, but (in contrast) not all TS is valid JS (before compilation).

I rewrote the existing type declaration to TS, you will notice that they are really straight forward, thus this relatively trivial usage need not be explained here. If you want a quick introduction into the TS syntax, read the [TS quickstart guide](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html).  

The tooling I introduced in af609a9eb22520e928e3f0bddb9d0390df950ed2 builds an ESM (allowing [tree shaking](https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking)) and a CJS distribution version of the library, using `npm run build` (run this every time before publishing). Both will be published to npm. Depending on the features of the environment using this library, one of the two versions will be executed. See TLDR for usage (for production, development and testing).

### What do the new files do?

* **package.json**
  * The scripts have been explained. 
  * The added dev dependencies serve them. 
  * The `types` key links to the type declaration file (built as part of the build procedure `npm run build`)
  * The `exports` key declares which env should use which executable.
* **repl/src/repl.ts** - The entry for the repl/dev environment.
* **index.html** - Used for in `npm run dev`; Is the index page to dev webserver serves
* **rollup.node.common.config.js** & **rollup.node.dev.config.js** - Build settings for repl for node execution
*  **rollup.web.common.config.js** & **rollup.web.dev.config.js** - Build settings for repl web execution
* **simpleWebServer.js** - Simple web server for web repl
* **tsconfig.dev.json** - Part of the build ts configuration for development
*  **tsconfig.json** - ts intellesense configuration (especially for vscode)
* **tsconfig.prod.cjs.json** - Production TS CJS build configuration
* **tsconfig.prod.esm.json** - Production TS ESM build configuration

### CODE REVIEW

**Comment by bminer** [comment]
I can tell you have spent a fair amount of time on this PR; for that I am grateful.  Surprisingly, this PR is essentially a full rewrite of this library.  It turns this from a JS project to a TypeScript project, and I'm not confident that it provides significant value to the community.  I'd almost lean toward publishing your code as a separate repo / package.

I'd love to discuss further.  What do you think?

**Comment by YChebotaev** [comment]
Done separately (sorry...)

**Comment by bminer** [comment]
@YChebotaev - Do you have a Git repo publicly available with your changes? I'd be happy to link to it from the README.

**Comment by bminer** [comment]
In that case, closing this...

**Comment by maximilianMairinger** [comment]
@YChebotaev And why exactly did you do this separately? Especially since now both ESM (see #12) as well as CJS (see #11) don't work? I could reproduce both errors.

**Comment by maximilianMairinger** [comment]
@bminer Sorry for the late answer. I am sure It provides value for the community, since it did to me. And as every non ts user is not impacted by this change, I don't see a reason to not go forward with this tbh. I know that it looks like a lot. Especially with all those new files, that's why I tried to break down the changes and explain the reason behind them. 

Note that I tried to keep the commits separate to improve readability. In essence [228bc04](https://github.com/YChebotaev/json-decycle/pull/10/commits/228bc04b834d9ec51458c1802df29652397a7c19) is the most scary change, as it migrates the (already existing) types to typescript syntax. 
The other commit changing the source file ([42b6bee](https://github.com/YChebotaev/json-decycle/pull/10/commits/42b6bee86c8db9aa77ad0e75b6fddd1858f4a632)) is trivial.

With the recent developments of this repo (mainly that it simply doesn't work anymore as of [547be0d](https://github.com/YChebotaev/json-decycle/commit/547be0d7dee797e1a96b0658bdf9196ed5873ede), but also that I disagree on #6) I feel inclined to publish my own version of this package on npm, assuming you are ok with that.

**Comment by bminer** [comment]
I think I will defer to @YChebotaev, since he did the TypeScript rewrite. Thoughts, @YChebotaev?

**Comment by Foxhoundn** [comment]
so - the issue is still ongoing - how to use this lib with TS? :)
